src/protocol/context/malicious.rs

```diff
@@ -256,6 +294,18 @@ impl<'a, F: Field> ContextInner<'a, F> {
         })
     }
 
+    #[must_use]
+    fn set_total_upgrades(&self, total_upgrades: usize) -> Arc<Self> {
```

Comment:
I have a comment later on that suggests that maybe this is a constructor argument on the validator itself (which is passed to the constructor of the malicious context, saving the need for a setter).